wallet, refactor: Remove Legacy wallet unused warnings and errors

pablomartin4btc opened the pull request:
Remove dead code due to legacy wallet support removal.

These changes have no impact on functionality. They are transparent to the end user, as legacy wallets can't be created or loaded anymore, so these checks are no longer reached. The legacy-to-descriptor wallet migration flow is not affected either, as these removals are not part of its process.

pablomartin4btc committed:
wallet, refactor: Remove Legacy warnings and errors

Remove dead code due to legacy wallet removal.

fanquake commented on src/qt/bitcoinstrings.cpp:
```diff
@@ -243,15 +243,6 @@ QT_TRANSLATE_NOOP("bitcoin-core", ""
-"future. Legacy wallets can be migrated to a descriptor wallet with "
-"migratewallet."),
-QT_TRANSLATE_NOOP("bitcoin-core", ""
 "Warning: Dumpfile wallet format \"%s\" does not match command line specified "
```
You can drop the changes in here, this file will be updated in a translations update. @hebasto might be worth doing one shortly just to cleanup these strings, since they keep being removed in various PRs.

pablomartin4btc committed:
wallet, refactor: Remove unused SetupGeneration

SetupGeneration was supposed to be the function that all SPKMs used
to setup automatic generation, but it didn't work out that way and
ended up being legacy only. It should be deleted at this point.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -382,11 +377,6 @@ std::shared_ptr<CWallet> CreateWallet(WalletContext& context, const std::string&
     const SecureString& passphrase = options.create_passphrase;
 
     if (wallet_creation_flags & WALLET_FLAG_DESCRIPTORS) options.require_format = DatabaseFormat::SQLITE;
-    else {
```
In e66e9ee9f19413c93d8523afdb8f9f3e07b818ec "wallet, refactor: Remove Legacy warnings and errors"

I'd prefer to at least preserve enforcement that the wallet being created is a descriptor wallet. Perhaps we could add an `Assert()` or `Assume()` for the flag.

pablomartin4btc replied:
I did that earlier but got a failure and I thought it was related with migration and rolled it back without checking it properly. I'll add it back.

pablomartin4btc commented:
_<ins>Updates</ins>_:
* Addressed @achow101's [feedback](https://github.com/bitcoin/bitcoin/pull/32481#discussion_r2087371964) by adding an `Assert()` on `CreateWallet()` enforcing only descriptor wallets can be created.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -381,12 +376,10 @@ std::shared_ptr<CWallet> CreateWallet(WalletContext& context, const std::string&
+    // Only descriptor wallets can be created
+    Assert(wallet_creation_flags & WALLET_FLAG_DESCRIPTORS);
+
     if (wallet_creation_flags & WALLET_FLAG_DESCRIPTORS) options.require_format = DatabaseFormat::SQLITE;
```
In 5111176aec91396320c7a337c0613389c1a5f4c0 "wallet, refactor: Remove Legacy warnings and errors"

This `if` is not needed anymore with the above `Assert`.

pablomartin4btc replied:
maybe neither do this one:

https://github.com/bitcoin/bitcoin/blob/39090518f62cca5432413d1669a85924ba721278/src/wallet/wallet.cpp#L448

(just to confirm and do both in 1 go)

achow101 replied:
Yes

pablomartin4btc commented:
Addressed [feedback](https://github.com/bitcoin/bitcoin/pull/32481#discussion_r2089806374) removing unnecessary descriptor wallet checks.

pablomartin4btc committed:
rpc, wallet, refactor: Remove non-descriptor errors

It is not possible to load a legacy/ non-descriptor wallet anymore
so no need to check for WALLET_FLAG_DESCRIPTORS in RPC calls, even when
passing -rpcwallet/ JSON `/wallet/<walletname>/` endpoint, that searches
for the wallets loaded already in the context.

pablomartin4btc commented:
_<ins>Updates</ins>_:
* Removed "non-descriptor" checks and error instances in RPC (noticed while reviewing #32708).

rkrux reviewed (commented):
Looking good based on a cursory glance, will take a deeper look soon.
Maybe mention a note in the PR description about wallet migration and how removing these checks don't affect that flow?

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -405,7 +397,7 @@ std::shared_ptr<CWallet> CreateWallet(WalletContext& context, const std::string&
 
     // Descriptor support must be enabled for an external signer wallet
-    if ((wallet_creation_flags & WALLET_FLAG_EXTERNAL_SIGNER) && !(wallet_creation_flags & WALLET_FLAG_DESCRIPTORS)) {
+    if (wallet_creation_flags & WALLET_FLAG_EXTERNAL_SIGNER) {
```
In 8fbb1c9657ef034aa179655a0869db45be2aaac8 "wallet, refactor: Remove Legacy warnings and errors"

That's wrong and causing a CI failure. I think this check is unneeded.

pablomartin4btc replied:
Yeah, my bad, also de error description was clear. Just found another case, I'll fix this soon. Thanks!

pablomartin4btc replied:
Sorry, having said that, the error description needs to be updated mentioning that external signer should be enabled instead of descriptor wallet.

pablomartin4btc commented:
_<ins>Updates</ins>_:
* Addressed [feedback](https://github.com/bitcoin/bitcoin/pull/32481#discussion_r2138353475) from @maflcko [and](https://github.com/bitcoin/bitcoin/pull/32481#discussion_r2138470337) @achow101.
* Removed also wallet descriptor check from `CWallet::EncryptWallet()` and  `CWallet::Create()`.

rkrux reviewed (commented):
utACK ce90f0c99fded22dd24f08757d6f48b5c6b52990

Looks fine to me, I navigated through the codebase to verify the diff's correctness. Suggested one more dead code removal if makes sense.

pablomartin4btc commented on src/wallet/wallet.cpp:
Yup, I was already planning to do in a separate PR since I found an issue/ improvement in `importdescriptors` RPC.

pablomartin4btc commented on src/wallet/wallet.cpp:
```diff
@@ -288,11 +288,6 @@ std::shared_ptr<CWallet> LoadWalletInternal(WalletContext& context, const std::s
             return nullptr;
         }
 
-        // Legacy wallets are being deprecated, warn if the loaded wallet is legacy
-        if (!wallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) {
-            warnings.emplace_back(_("Wallet loaded successfully. The legacy wallet type is being deprecated and support for creating and opening legacy wallets will be removed in the future. Legacy wallets can be migrated to a descriptor wallet with migratewallet."));
-        }
```
Mmm... I doubt it... just double checked running `WalletMigrationTest.test_encrypted()` (which would produced the unlock failure, similar behaviour as the backup failure, since both would `return util::Error`) with the removed check (-291 to -294) and that didn't reach the un-removed lines.

pablomartin4btc replied:
Right, there are 2 different things (or flows within `LoadWalletInternal`). First what I tried to explain in my previous comment (at least in my head) was that the migration process won't reach those lines (inside the `if` condition) because when that flow goes thru `LoadWalletInternal` it does it with the purpose of creating the target descriptor wallet so at that check/ condition the wallet is always a descriptor already. You can execute the migration using the old node (`self.old_node.migratewallet`) and will be the same thing. The second point is the "loading" behaviour of `LoadWalletInternal`, in order to hit it you had to call `loadwallet` RPC for a legacy wallet using the old node (v28 node or earlier), a legacy can't be loaded on the master node anymore (a validation cuts the flow earlier), that's why it's dead code (you can see that the test validating that warning was removed from `test/functional/wallet_backwards_compatibility.py` in #31250). Maybe you were saying the same thing but just wanted to clarify this for other ppl that could ever read this. Having `LoadWalletIternal` these 2 different behaviours (loading and creating) seems incorrect and brings confusion, I see there's already some interest in order to fix this in #32636. Thanks for your review!

achow101 commented:
ACK ce90f0c99fded22dd24f08757d6f48b5c6b52990
